skytemple_files/_resources/patches/asm_patches/cipnit_asm_mods/completeteamcontrol.asm

```diff
@@ -0,0 +1,299 @@
+	push r14
+	push r0-r1
+	push r5-r7
+	ldr r5,=NA_0235355C		;pointer to pokemon currently in control
+	sub r6,r5,NA_0235355C -NA_02353538
+	ldr r6,[r6]
+	add r6,12000h
+	ldr r6,[r6,0B28h]		;first pokemon's address
+	ldr r7,[r5]				;address of currently controled pokemon
+	ldr r0,=@CurrentLeader	;temporarally storing current pokemon's MDA here
+	str r7,[r0]
+	sub r0,r0,@CurrentLeader -@ManualModeOn
+	ldrb r1,[r0]
+	cmp r1,0h		;if it isn't manual mode, leaders aren't switched. If the player decides to change leaders via the hidden option in the team submenu, then without this check, the first pokemon would be made the leader every time the new leader moves. This is a problem because, if the new leader dies on its turn from a self-damaging attack, and my code changed the leader to another pokemon, the game will bug up.
+	bne @@leaderprepper_everythingsfine
+	pop r5-r7
+	pop r0-r1
+	bl NA_022FE4BC
+	pop r15
+@@leaderprepper_everythingsfine:
+	bl @leaderchanger
+	pop r5-r7
+	pop r0-r1
+	bl NA_022FE4BC	;this function executes the lead Pokemon's action. It does not wait for player input, it just executes the action that the player chose earlier, in a prior function which waits for player input.
+	push r0-r1
+	push r5-r7
+	ldr r5,=NA_0235355C
+	ldr r7,[r5]
+	cmp r7,0h				;if this is 0, leader is die forever
+	beq @@leaderprepper_end
+	ldr r0,=@CurrentLeader
+	ldr r6,[r0]
+	bl @leaderchanger
+@@leaderprepper_end:
+	pop r0-r1
+	pop r5-r7
+	pop r15
+.pool
+
+@paranoia_agent:	;This prevents the player from sending pokemon home during manual mode, which causes the game to bug out if you send the leader home, and from changing leaders using the hidden option during manual mode. 
+	push r14
+	push r0
+	cmp r1,34h		;farewell decision
+	cmpne r1,3Bh	;change leader decision
+	bne @@paranoiaagent_nvm
+	ldr r0,=@ManualModeOn
+	ldrb r0,[r0]
+	cmp r0,1h
+	bne @@paranoiaagent_nvm
+	mov r1,0h
+@@paranoiaagent_nvm:
+	pop r0
+	bl NA_022EB408	;this records the player's decision in the team submenu
+	pop r15
+.pool
+
+@WaitChecker:		;This skips a pokemon's turn if the pokemon swapped places with another pokemon previously. In the original game, swapping places with a partner skips the partner's next turn, as if the partner "used its move" when it swapped places with you. (A bug in the game raises this skip flag for the leader too, which I nopped earlier.) The game has nothing to prevent a leader from moving if this flag is raised, so this exists to make the game play as it's supposed to.
+	push r1-r3,r14
+	ldrsh r0,[r0,r1]
+	cmp r0,0h
+	popeq r1-r3,r15		;If this is equal to 0, the Pokemon wasn't going to take a turn anyway
+	ldr r1,[r4,0B4h]	;r4 has the current pokemon's entity address
+	ldrh r2,[r1]
+	mov r3,4000h
+	and r3,r2,r3
+	cmp r3,0h

```

Comment:
```suggestion
	tst r2,4000h
```

Reply:
Removing the `mov` operation is a problem - r3 is later used to subtract r2, so this leads to undefined behavior. I'll replace `and` and `cmp` though.